subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/TestTaskJvmArgsProviderIntegrationTest.groovy

```diff
@@ -80,4 +130,126 @@ class TestTaskJvmArgsProviderIntegrationTest extends AbstractIntegrationSpec {
+            file('file.txt').text = "Test"
+        } : file('original-file.txt')
+        def moved = isDirectory ? file('moved-directory').create {
+            file('file.txt').text = "Test"
+        } : file('other-location.txt')
+        def changed = isDirectory ? file('changed-directory').create {
+            file('file.txt').text = "changed"
+        } : file('changed-file.txt')
+        file('original-file.txt').text = "Test"
+        file('other-location.txt').text = "Test"
+        file('changed-file.txt').text = "changed"
+
+        when:
+        run 'test', "-PfilePath=${original}"
+        then:
+        executedAndNotSkipped ':test'
+
+        when:
+        run 'test', "-PfilePath=${moved}"
+        then:
+        skipped ':test'
+
+        when:
+        fails 'test', "-PfilePath=${changed}"
+        then:
+        executedAndNotSkipped ':test'
+        failure.assertHasCause("There were failing tests.")
+
+        where:
+        systemPropertyName     | value                      | isDirectory
+        'input.file.path'      | 'inputFile(filePath)'      | false
+        'input.directory.path' | 'inputDirectory(filePath)' | true
+    }
+
+    @Unroll
+    def "can pass outputs as system properties via #value"() {
+        buildFile << """
+            test.systemProperties {
+                add '${systemPropertyName}', ${value}
+            }
+        """
+
+        def output = isDirectory ? file('directory') : file('output-file.txt')
+
+        def outputFile = isDirectory ? output.file('file.txt') : output
+
+        when:
+        run 'test', "-PfilePath=${output}"
+        then:
+        executedAndNotSkipped ':test'
+        outputFile.text == 'Test'
+
+        when:
+        run 'test', "-PfilePath=${output}"
+        then:
+        skipped ':test'
+
+        when:
+        outputFile.text = "changed"
+        run 'test', "-PfilePath=${output}"
+        then:
+        executedAndNotSkipped ':test'
+        outputFile.text == 'Test'
+
+        where:
+        systemPropertyName      | value                       | isDirectory
+        'output.file.path'      | 'outputFile(filePath)'      | false
+        'output.directory.path' | 'outputDirectory(filePath)' | true
+    }
+
+    @Unroll
+    def "can pass #type as ignored system property"() {
+        buildFile << """
+            test.systemProperties {
+                add 'path.ignored', ignored(${expression})
+            }
+        """
+
+        when:
+        run 'test', "-PpropertyValue=${value}"
+        then:
+        executedAndNotSkipped ':test'
+
+        when:
+        run 'test', "-PpropertyValue=${changedValue}"
+        then:
+        skipped ':test'
+
+        where:
+        type | expression | value | changedValue
+        'file'   | 'file(propertyValue)'            | 'ignored.txt' | 'ignored2.text'
+        'object' | 'Integer.valueOf(propertyValue)' | '25'          | '38'
+    }
+
+    def "can pass providers as system properties"() {
```

Comment:
If we allow this we should test if the build dependencies are retained.

Reply:
Keeping the dependencies will only work after https://github.com/gradle/build-cache/issues/987 has been merged.

Reply:
I added a `ToBeImplemented` test.